src/sentry/workflow_engine/processors/detector.py

```diff
@@ -55,6 +57,97 @@ def get_detector_by_event(event_data: WorkflowEventData) -> Detector:
+    if not event_list:
+        return {}
+
+    # Separate events by whether they have occurrences or not
+    events_with_occurrences: list[tuple[GroupEvent, IssueOccurrence]] = []
+    events_without_occurrences: list[GroupEvent] = []
+
+    for event in event_list:
+        issue_occurrence = event.occurrence
+
+        if issue_occurrence is None:
+            events_without_occurrences.append(event)
+        else:
+            events_with_occurrences.append((event, issue_occurrence))
+
+    result: dict[str, Detector] = {}
+
+    # Fetch detectors for events with occurrences (by detector_id)
+    if events_with_occurrences:
+        detector_ids = []
+        event_to_detector_id: dict[GroupEvent, int] = {}
+
+        for event, occurrence in events_with_occurrences:
+            detector_id = occurrence.evidence_data.get("detector_id")
+            if detector_id is not None:
+                detector_ids.append(detector_id)
+                event_to_detector_id[event] = detector_id
+
+        if detector_ids:
+            detectors_by_id = {
+                detector.id: detector for detector in Detector.objects.filter(id__in=detector_ids)
+            }
+
+            for event, detector_id in event_to_detector_id.items():
+                detector = detectors_by_id.get(detector_id)
+                if detector:
+                    result[event.event_id] = detector
+                else:
+                    metrics.incr("workflow_engine.detectors.error")
+                    logger.warning(
+                        "Detector not found for event",
+                        extra={
+                            "event_id": event.event_id,
+                            "group_id": event.group_id,
+                            "detector_id": detector_id,
+                        },
+                    )
+
+    # Fetch detectors for events without occurrences (by project_id and issue_type)
+    if events_without_occurrences:
+        # Group events by project_id
+        project_events: dict[int, list[GroupEvent]] = defaultdict(list)
+        detector_types: set[str] = set()
+
+        for event in events_without_occurrences:
+            project_events[event.project_id].append(event)
+            detector_types.add(event.group.issue_type.slug)
+
+        # Fetch all detectors for the projects in one query
+        project_ids = list(project_events.keys())
+        detectors_by_project: dict[int, dict[str, Detector]] = defaultdict(dict)
+
+        for detector in Detector.objects.filter(
+            project_id__in=project_ids, type__in=detector_types
+        ):
+            detectors_by_project[detector.project_id][detector.type] = detector
+
+        # Assign detectors to events
+        for project_id, events in project_events.items():
+            project_detectors = detectors_by_project.get(project_id, {})
+            for event in events:
+                detector = project_detectors.get(event.group.issue_type.slug)
```

Comment:
the `detector_by_project` and `project_detectors` is a bit confusing when trying to read it. do they need to be separate?